### PR TITLE
Add new Gatsby plugin for Inferno support

### PR DESCRIPTION
**Pull request by DonnieWest**
Created as a proof of concept for #8858

You can see others who have had success with getting Inferno V6 running on Gatsby and I've also had some luck. Figured I'd essentially fork the Preact plugin and open this as a place for discussion. I'm willing to make any edits requested :)

### CODE REVIEW

File: packages/gatsby-plugin-inferno/package.json [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  "homepage": "https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-plugin-inferno#readme",
+  "keywords": [
+    "gatsby",
+    "gatsby-plugin",
+    "inferno"
+  ],
+  "license": "MIT",
+  "main": "index.js",
+  "peerDependencies": {
+    "gatsby": ">2.0.0-alpha"
```
Comment (pieh) [inline packages/gatsby-plugin-inferno/package.json]: We would need `inferno` and  `inferno-compat` as peerDeps too

File: packages/gatsby-plugin-inferno/.gitignore [hunk 2]
```diff
@@ -0,0 +1,2 @@
+/gatsby-node.js
+/gatsby-browser.js
```
Comment (pieh) [inline packages/gatsby-plugin-inferno/.gitignore]: I think it should be something like:
```
/*.js
!index.js
yarn.lock
```
We don't want any `gatsby-` built file in git (including `gatsby-node`)

File: packages/gatsby-plugin-inferno/src/gatsby-node.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
+exports.onCreateWebpackConfig = ({ stage, actions }) => {
+  // Requiring the server version of React-dom is hardcoded right now
+  // in the development server. So we'll just avoid loading Inferno there
```
Comment (pieh) [inline packages/gatsby-plugin-inferno/src/gatsby-node.js]: Was this copied from `preact` plugin or this is something new? I'm not sure what's the problem this check is workarounding

**Comment by DonnieWest** [comment]
@simonjoom I'm getting around the hot reloading issue by just excluding Inferno during development like the Preact plugin does. It's definitely not an ideal solution, but I'd prefer to maintain consistency with how other plugins do things unless strictly necessary.

This was all I personally needed to do in order to switch to Inferno and seems pretty consistent with how [their docs tell you to migrate](https://github.com/infernojs/inferno/tree/master/packages/inferno-compat#usage-with-webpack) - unless I'm missing something, we shouldn't have to do anything else. For instance, switching the pragma isn't needed because Inferno provides `React.createElement`, right?

File: packages/gatsby-plugin-inferno/.gitignore (hunk 2 above)
Comment (DonnieWest) [reply packages/gatsby-plugin-inferno/.gitignore]: I'll try to get to this tonight - thanks :)

File: packages/gatsby-plugin-inferno/package.json (hunk 1 above)
Comment (DonnieWest) [reply packages/gatsby-plugin-inferno/package.json]: Would something like this work?

```
  "peerDependencies": {
    "inferno-compat": "6.x",
    "gatsby": ">2.0.0-alpha"
},
```

It's only Inferno 6 that works at this point

File: packages/gatsby-plugin-inferno/src/gatsby-node.js (hunk 3 above)
Comment (pieh) [reply packages/gatsby-plugin-inferno/src/gatsby-node.js]: `develop-html` is only for generating things for basic `index.html` used in develop mode. There is separate `develop` stage that actually produce js bundle, so not sure if this changes anything for hot-reload

**Comment by pieh** [comment]
Just a heads up - `preact` plugin has some problems currently https://github.com/gatsbyjs/gatsby/issues/8545, so I wouldn't take how that one is doing as sure thing forward.

**Comment by simonjoom** [comment]
> @simonjoom I'm getting around the hot reloading issue by just excluding Inferno during development like the Preact plugin does. It's definitely not an ideal solution, but I'd prefer to maintain consistency with how other plugins do things unless strictly necessary.
> 
> This was all I personally needed to do in order to switch to Inferno and seems pretty consistent with how [their docs tell you to migrate](https://github.com/infernojs/inferno/tree/master/packages/inferno-compat#usage-with-webpack) - unless I'm missing something, we shouldn't have to do anything else. For instance, switching the pragma isn't needed because Inferno provides `React.createElement`, right?
> 
> Edit: Maybe I'm not actually fixing the hot reloading issue. I'll verify tonight and get back

Hello
About development i think it s good to have an inferno development structure without hotreload

First because it's avoid you to deploy everytime your project to be able to reduce bug involving from inferno. 
Second to have a development more near "production" but keeping the HMR and sourcemap that is great. As well have a development mode with no react working (only inferno) is great to test some library like react-transition-group (It's working even with inferno)

so for me it's why i create a new  variable process.env.NODE_ENV=="inferno" 

In my case i use
" npm run develop" who give me exactly the same than before hacking with gastby 

"npm run inferno" give me my test production case  so process.env.NODE_ENV=="inferno"  here (hot reloaded avoided with this boolean) HMR working and no minifyjs

"npm run build" produce output with inferno inside without react + minifyjs

**Comment by DonnieWest** [comment]
@simonjoom as mentioned, I agree it's not the ideal situation. This at least serves my personal use case and I don't think the loss of HMR is worth it in development

I'll let someone else (preferably from the Gatsby team) decide which route to go

**Comment by pieh** [comment]
What is the problem with inferno and HMR - it straight doesn't work or it's buggy?

**Comment by simonjoom** [comment]
> What is the problem with inferno and HMR - it straight doesn't work or it's buggy?

With inferno:
HMR is working 
the webpack-hot-middleware i mean work well and refresh the page after mods 

But it's the react-hot-loader who cannot work with. because react-hot-loader was done for react and still they did not release a inferno-hot-loader to replace it

**Comment by pieh** [comment]
Ah, that makes sense - yeah then we should just refresh page in there is no equiavalent to `react-hot-loader` for inferno. But definitely should use inferno for development to keep it as close to production as possible

**Comment by DonnieWest** [comment]
@simonjoom @pieh sounds good then. I'll release the code here as my own package since this is all I need and let @simonjoom take this effort from here :)

**Comment by wardpeet** [comment]
@DonnieWest @simonjoom I'll close this one as if I read correct @DonnieWest is not going to pursuit this on the gatsby repo. @simonjoom  feel free to create a new PR with your own code.

You can also just publish it on your own and add it to our lovely plugin directory.
https://www.gatsbyjs.org/docs/submit-to-plugin-library/

**Pull request closed without merging**